File: yara.go
package ozma

import (
	"embed"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
	"unicode/utf8"
)

//go:embed src/doc_analyse/detection/default.yara
var yaraAssets embed.FS

const DefaultYaraRulesFile = "default.yara"

type YaraGlossaryError struct{ Message string }

func (e YaraGlossaryError) Error() string { return e.Message }

type compiledPattern struct {
	Name    string
	Regexp  *regexp.Regexp
	Pattern string
}

type yaraRuleMeta struct {
	RuleID                string
	Category              string
	Severity              string
	Weight                float64
	RouteHint             string
	RequiresLLMValidation bool
	Reason                string
}

type compiledRule struct {
	Name      string
	Meta      yaraRuleMeta
	Patterns  []compiledPattern
	Condition string
}

type YaraDetector struct {
	compiled []compiledRule
}

func NewYaraDetector() (*YaraDetector, error) {
	source, err := yaraAssets.ReadFile("src/doc_analyse/detection/default.yara")
	if err != nil {
		return nil, YaraGlossaryError{Message: "Default YARA rules failed to load."}
	}
	compiled, err := CompileYaraRules(string(source))
	if err != nil {
		return nil, err
	}
	return &YaraDetector{compiled: compiled}, nil
}

func YaraDetectorFromFile(path string) (*YaraDetector, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, YaraGlossaryError{Message: fmt.Sprintf("Could not read YARA rules file '%s': %v", path, err)}
	}
	compiled, err := CompileYaraRules(string(b))
	if err != nil {
		return nil, err
	}
	return &YaraDetector{compiled: compiled}, nil
}

func CompileYaraRules(source string) ([]compiledRule, error) {
	blocks := splitRuleBlocks(source)
	rules := []compiledRule{}
	for _, block := range blocks {
		rule, err := parseRuleBlock(block)
		if err != nil {
			return nil, err
		}
		if rule.Name != "" {
			rules = append(rules, rule)
		}
	}
	return rules, nil
}

func (d *YaraDetector) Detect(chunk TextChunk) ([]DetectionFinding, error) {
	if strings.TrimSpace(chunk.Text) == "" {
		return nil, nil
	}
	if d == nil || len(d.compiled) == 0 {
		return nil, YaraGlossaryError{Message: "Default YARA rules failed to load. Supply custom rules via YaraDetectorFromFile()."}
	}
	byteToChar, _ := chunk.Metadata["byte_to_char"].([]int)
	if byteToChar == nil {
		byteToChar = BuildByteToChar(chunk.Text)
	}
	findings := []DetectionFinding{}
	for _, rule := range d.compiled {
		matchesByName := map[string][]regexpMatch{}
		for _, pattern := range rule.Patterns {
			for _, loc := range pattern.Regexp.FindAllStringIndex(chunk.Text, -1) {
				span := chunk.Text[loc[0]:loc[1]]
				if strings.TrimSpace(span) == "" {
					continue
				}
				matchesByName[pattern.Name] = append(matchesByName[pattern.Name], regexpMatch{span: span, startByte: loc[0]})
			}
		}
		if !ruleConditionMatches(rule, matchesByName) {
			continue
		}
		for _, matches := range matchesByName {
			for _, match := range matches {
				charOffset := match.startByte
				if match.startByte >= 0 && match.startByte < len(byteToChar) {
					charOffset = byteToChar[match.startByte]
				}
				score := (*float64)(nil)
				if rule.Meta.Weight > 0 {
					v := rule.Meta.Weight / 100
					if v > 1 {
						v = 1
					}
					score = &v
				}
				findings = append(findings, BuildFinding(
					chunk,
					match.span,
					rule.Meta.Category,
					rule.Meta.Severity,
					rule.Meta.Reason,
					rule.Meta.RuleID,
					chunk.StartChar+charOffset,
					chunk.StartChar+charOffset+utf8.RuneCountInString(match.span),
					score,
					rule.Meta.RequiresLLMValidation,
					Metadata{
						"detector":    "YaraDetector",
						"yara_rule":   rule.Meta.RuleID,
						"yara_weight": rule.Meta.Weight,
						"route_hint":  rule.Meta.RouteHint,
					},
				))
			}
		}
	}
	return finalizeFindings(findings), nil
}

type regexpMatch struct {
	span      string
	startByte int
}

func splitRuleBlocks(source string) []string {
	re := regexp.MustCompile(`\brule\s+\w+\s*\{`)
	locs := re.FindAllStringIndex(source, -1)
	blocks := []string{}
	for i, loc := range locs {
		end := len(source)
		if i+1 < len(locs) {
			end = locs[i+1][0]
		}
		blocks = append(blocks, source[loc[0]:end])
	}
	return blocks
}

func parseRuleBlock(block string) (compiledRule, error) {
	nameMatch := regexp.MustCompile(`\brule\s+(\w+)\s*\{`).FindStringSubmatch(block)
	if len(nameMatch) < 2 {
		return compiledRule{}, nil
	}
	name := nameMatch[1]
	meta := parseYaraMeta(sectionBetween(block, "meta:", "strings:"), name)
	patterns, err := parseYaraStrings(sectionBetween(block, "strings:", "condition:"))
	if err != nil {
		return compiledRule{}, err
	}
	condition := strings.TrimSuffix(strings.TrimSpace(sectionUntilClose(block, "condition:")), "}")
	return compiledRule{Name: name, Meta: meta, Patterns: patterns, Condition: condition}, nil
}

func sectionBetween(block, startMarker, endMarker string) string {
	start := strings.Index(block, startMarker)
	if start < 0 {
		return ""
	}
	start += len(startMarker)
	end := strings.Index(block[start:], endMarker)
	if end < 0 {
		return block[start:]
	}
	return block[start : start+end]
}

func sectionUntilClose(block, startMarker string) string {
	start := strings.Index(block, startMarker)
	if start < 0 {
		return ""
	}
	return block[start+len(startMarker):]
}

func parseYaraMeta(section, ruleName string) yaraRuleMeta {
	values := map[string]string{}
	lineRE := regexp.MustCompile(`(?m)^\s*([A-Za-z_]\w*)\s*=\s*(.+?)\s*$`)
	for _, m := range lineRE.FindAllStringSubmatch(section, -1) {
		values[strings.ToLower(m[1])] = strings.TrimSpace(m[2])
	}
	get := func(key, fallback string) string {
		raw := values[strings.ToLower(key)]
		if raw == "" {
			return fallback
		}
		if strings.HasPrefix(raw, `"`) && strings.HasSuffix(raw, `"`) {
			return strings.Trim(raw, `"`)
		}
		return raw
	}
	weight, _ := strconv.ParseFloat(get("weight", "0"), 64)
	return yaraRuleMeta{
		RuleID:                get("rule_id", ruleName),
		Category:              get("category", "unknown"),
		Severity:              get("severity", "medium"),
		Weight:                weight,
		RouteHint:             get("route_hint", "evidence"),
		RequiresLLMValidation: parseBool(get("requires_llm_validation", "false")),
		Reason:                get("reason", "YARA rule matched."),
	}
}

func parseYaraStrings(section string) ([]compiledPattern, error) {
	stringRE := regexp.MustCompile(`(?is)\$([A-Za-z_]\w*)\s*=\s*(/(?:[^/\\]|\\.)*/[A-Za-z]*|"(?:[^"\\]|\\.)*")(?:\s+nocase)?`)
	patterns := []compiledPattern{}
	for _, m := range stringRE.FindAllStringSubmatch(section, -1) {
		name, raw, full := m[1], m[2], strings.ToLower(m[0])
		nocase := strings.Contains(full, "nocase")
		var expr string
		if strings.HasPrefix(raw, "/") {
			last := strings.LastIndex(raw, "/")
			expr = raw[1:last]
			flags := raw[last+1:]
			nocase = nocase || strings.Contains(flags, "i")
		} else {
			unquoted, err := strconv.Unquote(raw)
			if err != nil {
				unquoted = strings.Trim(raw, `"`)
			}
			expr = regexp.QuoteMeta(unquoted)
		}
		if nocase {
			expr = "(?i)" + expr
		}
		re, err := regexp.Compile(expr)
		if err != nil {
			return nil, YaraGlossaryError{Message: fmt.Sprintf("Failed to compile YARA pattern %s: %v", name, err)}
		}
		patterns = append(patterns, compiledPattern{Name: name, Regexp: re, Pattern: raw})
	}
	return patterns, nil
}

func ruleConditionMatches(rule compiledRule, matches map[string][]regexpMatch) bool {
	condition := strings.ToLower(rule.Condition)
	if strings.Contains(condition, "any of them") {
		return len(matches) > 0
	}
	if strings.Contains(condition, "all of them") {
		return len(matches) >= len(rule.Patterns)
	}
	for _, pattern := range rule.Patterns {
		condition = strings.ReplaceAll(condition, "$"+strings.ToLower(pattern.Name), fmt.Sprintf("%t", len(matches[pattern.Name]) > 0))
	}
	condition = strings.ReplaceAll(condition, "(", " ")
	condition = strings.ReplaceAll(condition, ")", " ")
	orParts := strings.Split(condition, " or ")
	for _, orPart := range orParts {
		andParts := strings.Split(orPart, " and ")
		ok := true
		sawBool := false
		for _, part := range andParts {
			part = strings.TrimSpace(part)
			if part == "true" {
				sawBool = true
				continue
			}
			if part == "false" {
				sawBool = true
				ok = false
			}
		}
		if sawBool && ok {
			return true
		}
	}
	return len(matches) > 0 && !strings.Contains(condition, "true") && !strings.Contains(condition, "false")
}

func parseBool(raw string) bool {
	switch strings.ToLower(strings.TrimSpace(raw)) {
	case "true", "yes", "1", "on":
		return true
	default:
		return false
	}
}

File: src/detection/yara.ts
/**
 * YARA detector with native RegExp reimplementation.
 * Ported from doc_analyse/detection/yara.py — replaces yara-python with JS RegExp.
 *
 * The YARA-to-RegExp transpiler parses the .yara source and converts each rule's
 * string patterns to JavaScript RegExp objects. The detector then runs all
 * compiled RegExp patterns against chunk text, producing DetectionFinding objects.
 */

import { readFileSync } from "fs";
import { DetectionFinding, createDetectionFinding } from "./models.js";
import { TextChunk } from "../ingestion/models.js";
import { _build_byte_to_char } from "../ingestion/chunking.js";
import { BaseDetector } from "./base.js";

// ---------------------------------------------------------------------------
// YARA rule types (internal)
// ---------------------------------------------------------------------------

interface YaraRuleMeta {
  rule_id: string;
  category: string;
  severity: string;
  weight: number;
  route_hint: string;
  requires_llm_validation: boolean;
  reason: string;
}

interface CompiledPattern {
  regex: RegExp;
  pattern: string;
}

interface CompiledRule {
  name: string;
  meta: YaraRuleMeta;
  patterns: CompiledPattern[];
  condition: "any" | "all";
}

export class YaraGlossaryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "YaraGlossaryError";
  }
}

// ---------------------------------------------------------------------------
// YARA source parser and RegExp transpiler
// ---------------------------------------------------------------------------

function _read_yara_source(filePath: string): string {
  try {
    return readFileSync(filePath, "utf-8");
  } catch (err) {
    throw new YaraGlossaryError(
      `Could not read YARA rules file '${filePath}': ${err}`,
    );
  }
}

function _read_packaged_yara_source(filename: string): string {
  const fileUrl = new URL(`./${filename}`, import.meta.url);
  return readFileSync(fileUrl, "utf-8");
}

export function compile_yara_rules(source: string): CompiledRule[] {
  const rules: CompiledRule[] = [];
  const ruleBlocks = _split_rule_blocks(source);

  for (const block of ruleBlocks) {
    const rule = _parse_rule_block(block);
    if (rule) rules.push(rule);
  }

  return rules;
}

function _split_rule_blocks(source: string): string[] {
  const blocks: string[] = [];
  const regex = /\brule\s+\w+\s*\{/g;
  const matches: { offset: number; line: number }[] = [];

  let match: RegExpExecArray | null;
  const lines = source.split("\n");
  let charOffset = 0;
  const lineToChar: number[] = [0];
  for (const line of lines) {
    charOffset += line.length + 1;
    lineToChar.push(charOffset);
  }

  while ((match = regex.exec(source)) !== null) {
    matches.push({
      offset: match.index,
      line: source.substring(0, match.index).split("\n").length,
    });
  }

  for (let i = 0; i < matches.length; i++) {
    const start = matches[i]!.offset;
    const end = i < matches.length - 1 ? matches[i + 1]!.offset : source.length;
    blocks.push(source.slice(start, end));
  }

  return blocks;
}

function _parse_rule_block(block: string): CompiledRule | null {
  const nameMatch = block.match(/\brule\s+(\w+)\s*\{/);
  if (!nameMatch) return null;
  const name = nameMatch[1]!;

  const metaMatch = block.match(/meta:\s*([\s\S]*?)(?=strings:)/);
  const stringsMatch = block.match(/strings:\s*([\s\S]*?)(?=condition:)/);
  const conditionMatch = block.match(/condition:\s*([\s\S]*?)\}/);

  const meta = _parse_meta(metaMatch?.[1] ?? "");
  const patterns = _parse_strings(stringsMatch?.[1] ?? "");
  const condition = _parse_condition(conditionMatch?.[1] ?? "");

  return { name, meta, patterns, condition };
}

function _parse_meta(metaSection: string): YaraRuleMeta {
  const values = new Map<string, string>();
  const metaLineRegex = /^\s*([a-zA-Z_]\w*)\s*=\s*(.+?)\s*$/gm;

  let match: RegExpExecArray | null;
  while ((match = metaLineRegex.exec(metaSection)) !== null) {
    values.set(match[1]!.toLowerCase(), match[2]!.trim());
  }

  const get = (key: string, fallback: string | number | boolean): string | number | boolean => {
    const rawValue = values.get(key.toLowerCase());
    if (rawValue === undefined) {
      return fallback;
    }

    if (rawValue.startsWith("\"") && rawValue.endsWith("\"")) {
      return rawValue.slice(1, -1);
    }

    if (/^(true|yes|1|on)$/i.test(rawValue)) {
      return true;
    }

    if (/^(false|no|0|off)$/i.test(rawValue)) {
      return false;
    }

    if (/^\d+\.?\d*$/.test(rawValue)) {
      return parseFloat(rawValue);
    }

    return rawValue;
  };

  return {
    rule_id: String(get("rule_id", nameFromMeta(metaSection) || "unknown")),
    category: String(get("category", "unknown")),
    severity: String(get("severity", "medium")),
    weight: Number(get("weight", 0)),
    route_hint: String(get("route_hint", "evidence")),
    requires_llm_validation: Boolean(get("requires_llm_validation", false)),
    reason: String(get("reason", "YARA rule matched.")),
  };
}

function nameFromMeta(_metaSection: string): string | null {
  return null;
}

function _parse_strings(stringsSection: string): CompiledPattern[] {
  const patterns: CompiledPattern[] = [];
  // Pattern: $var = /regex/flags  OR  $var = "literal"
  // Pattern is followed by optional nocase modifier.
  // Group 1: var name, Group 2: /regex body/flags OR "literal body"
  const stringRegex = /\$([a-zA-Z_]\w*)\s*=\s*(\/(?:[^\/\\]|\\.)*\/[gimsuvy]*|"(?:[^"\\]|\\.)*")(?:\s+nocase)?/gi;

  let match: RegExpExecArray | null;
  while ((match = stringRegex.exec(stringsSection)) !== null) {
    const rawPattern = match[2]!;
    const isRegex = rawPattern.startsWith("/");
    const hasNocase = rawPattern.toLowerCase().includes("nocase") ||
      match[0]!.toLowerCase().includes("nocase");

    let regex: RegExp;
    try {
      if (isRegex) {
        const lastSlash = rawPattern.lastIndexOf("/");
        const body = rawPattern.substring(1, lastSlash);
        const flags = new Set(rawPattern.substring(lastSlash + 1).split(""));
        flags.add("g");
        if (hasNocase) flags.add("i");
        regex = new RegExp(body, [...flags].join(""));
      } else {
        const escaped = rawPattern.slice(1, -1).replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
        regex = new RegExp(escaped, hasNocase ? "gi" : "g");
      }
    } catch {
      continue;
    }

    patterns.push({ regex, pattern: rawPattern });
  }

  return patterns;
}

function _parse_condition(condition: string): "any" | "all" {
  const trimmed = condition.trim();
  if (trimmed.includes("any of them")) return "any";
  if (trimmed.includes("all of them")) return "all";
  if (trimmed.includes(" or ")) return "any";
  if (trimmed.includes(" and ")) return "all";
  return "any";
}

// ---------------------------------------------------------------------------
// YaraDetector
// ---------------------------------------------------------------------------

export class YaraDetector extends BaseDetector {
  private readonly compiled: CompiledRule[];

  constructor(compiled?: CompiledRule[] | null) {
    super();
    this.compiled = compiled ?? _DEFAULT_COMPILED_RULES ?? [];
  }

  static from_file(filePath: string): YaraDetector {
    const source = _read_yara_source(filePath);
    const compiled = compile_yara_rules(source);
    return new YaraDetector(compiled);
  }

  detect(chunk: TextChunk): readonly DetectionFinding[] {
    if (!chunk.text || !chunk.text.trim()) return [];

    if (!this.compiled || this.compiled.length === 0) {
      throw new YaraGlossaryError(
        "Default YARA rules failed to load. Provide rules via from_file().",
      );
    }

    const byteToChar = (chunk.metadata["byte_to_char"] as readonly number[] | null)
      ?? _build_byte_to_char(chunk.text);

    const findings: DetectionFinding[] = [];

    for (const rule of this.compiled) {
      const ruleMatches: Array<{
        spanText: string;
        charOffset: number;
        spanCharLen: number;
      }> = [];

      for (const { regex } of rule.patterns) {
        regex.lastIndex = 0;
        let m: RegExpExecArray | null;
        while ((m = regex.exec(chunk.text)) !== null) {
          const spanText = m[0]!;
          if (!spanText.trim()) continue;

          const byteOffset = m.index;
          const charOffset = byteToChar[byteOffset] ?? byteOffset;
          const spanCharLen = spanText.length;

          ruleMatches.push({ spanText, charOffset, spanCharLen });

          if (!regex.global) break;
        }
      }

      if (rule.condition === "any") {
        if (ruleMatches.length === 0) continue;
      } else {
        if (ruleMatches.length < rule.patterns.length) continue;
      }

      for (const { spanText, charOffset, spanCharLen } of ruleMatches) {
        const normalizedScore =
          rule.meta.weight > 0 ? Math.min(1.0, rule.meta.weight / 100.0) : null;

        findings.push(
          createDetectionFinding({
            span: spanText,
            category: rule.meta.category,
            severity: rule.meta.severity,
            reason: rule.meta.reason,
            rule_id: rule.meta.rule_id,
            start_char: chunk.start_char + charOffset,
            end_char: chunk.start_char + charOffset + spanCharLen,
            source: chunk.source,
            requires_llm_validation: rule.meta.requires_llm_validation,
            score: normalizedScore,
            metadata: Object.freeze({
              detector: "YaraDetector",
              yara_rule: rule.name,
              yara_weight: rule.meta.weight,
              route_hint: rule.meta.route_hint,
            }),
          }),
        );
      }
    }

    return this._finalize_findings(findings);
  }
}

// ---------------------------------------------------------------------------
// Module-level default compiled rules
// ---------------------------------------------------------------------------

let _DEFAULT_COMPILED_RULES: CompiledRule[] | null = null;

export function _load_default_rules(): CompiledRule[] | null {
  try {
    const source = _read_packaged_yara_source("default.yara");
    return compile_yara_rules(source);
  } catch {
    return null;
  }
}

export function _init_default_rules(): void {
  _DEFAULT_COMPILED_RULES = _load_default_rules();
}

_init_default_rules();
